cpp/unreal_plugins/UrdfBot/Source/UrdfParser.cpp

```diff
@@ -0,0 +1,381 @@
+//
+// Copyright(c) 2022 Intel. Licensed under the MIT License <http://opensource.org/licenses/MIT>.
+//
+
+#include "UrdfParser.h"
+
+#include <XmlFile.h>
+#include <XmlNode.h>
+#include <XmlParser.h>
+
+#include "Assert/Assert.h"
+
+UrdfRobotDesc UrdfParser::parse(const std::string& file_name)
+{
+    FXmlFile file;
+    file.LoadFile(UTF8_TO_TCHAR(file_name.c_str()));
+
+    FXmlNode* robot_node = file.GetRootNode();
+    ASSERT(robot_node->GetTag().Equals(TEXT("robot")));
+
+    return parseRobotNode(robot_node);
+}
+
+UrdfRobotDesc UrdfParser::parseRobotNode(FXmlNode* robot_node)
+{
+    UrdfRobotDesc robot_desc;
+    // parse all nodes
+    for (auto& child_node : robot_node->GetChildrenNodes()) {
+        const FString& tag = child_node->GetTag();
+
+        if (tag.Equals(TEXT("link"))) {
+            UrdfLinkDesc link_desc = parseLinkNode(child_node);
+            ASSERT(!robot_desc.link_descs_.count(link_desc.name_));
+            robot_desc.link_descs_[link_desc.name_] = std::move(link_desc);
+        } else if (tag.Equals(TEXT("joint"))) {
+            UrdfJointDesc joint_desc = parseJointNode(child_node);
+            ASSERT(!robot_desc.joint_descs_.count(joint_desc.name_));
+            robot_desc.joint_descs_[joint_desc.name_] = std::move(joint_desc);
+        } else if (tag.Equals(TEXT("material"))) {
+            UrdfMaterialDesc material_desc = parseMaterialNode(child_node);
+            ASSERT(!robot_desc.material_descs_.count(material_desc.name_));
+            robot_desc.material_descs_[material_desc.name_] = std::move(material_desc);
+        }
+    }
+
+    // for a robot with N links, there must be N-1 joints.
+    ASSERT(robot_desc.joint_descs_.size() + 1 == robot_desc.link_descs_.size());
+
+    // build link tree based on joints
```

Comment:
```
// for each joint, update the joint's parent and child pointers, and the parent and child pointers for its links
```

Reply:
I think it is necessary to keep child_link's parent link/joint pointer to confirm the robot structure is correct, i.e. a link must have at most one parent link.

Reply:
Fair point. I agree that it is desirable to guarantee the correctness of the tree structure.

Rather than maintaining parent pointers, I like the idea of `ASSERT(!child_link_desc->has_parent_)` before setting it to `true` here. `has_parent_` would default to `false`, so if there is ever a child link with two parent links in a URDF file, this assert would catch it. Let's replace `is_root_` with `has_parent_` because it is a more clear name, and makes `ASSERT(!child_link_desc->has_parent_)` read more clearly.